### DISCUSSION TITLE
Ebyte E22-900M30S output power issues

### DISCUSSION THREAD

**Discussion by lucauriarte** [jgromes/RadioLib#487] [Q&A] [answered]
Hello, i have an [EBYTE E22-900M30S](https://www.ebyte.com/en/product-view-news.html?id=453) and i am not able to reach the 30dBm output power. According to [High power radio modules guide](https://github.com/jgromes/RadioLib/wiki/High-power-Radio-Modules-Guide) i have to set the output power to 4dBm to get a 30dBm output, because of the presence of an external 26dB PA. I didn´t find the PA part number on the product page or datasheet, but RadioLib wiki says it is an +26dB YP2233W.

I have measured the output power with a spectrum analyzer and i have found that the output power is 10dBm when it is set to 4dBm in code. So I wrote a code to test every allowed output power:

```
*****************************************************************************************************
  This code is used to test radio tx power. Use an external 5V power source to power the radio module
******************************************************************************************************/ 
#include <RadioLib.h>

#define RADIO_FREQUENCY 915.0     //Carrier frequency in MHz. 
#define RADIO_BANDWIDTH 125       //LoRa link bandwidth in kHz.
#define RADIO_SF 11               //LoRa link spreading factor.
#define RADIO_CODE_RATE 7         //LoRa link coding rate denominator.
#define RADIO_SYNC_WORD 0x12      //LoRa sync word. 
#define RADIO_OUTPUT_POWER 22     //Transmission output power in dBm. 
#define RADIO_PREAMBLE_LENGTH 8   //Length of LoRa transmission preamble in symbols. 
#define RADIO_LNA_RECIEVER_GAIN 0 //Gain of receiver LNA (low-noise amplifier). 

#define RADIO_CS   13
#define RADIO_DIO1 12 
#define RADIO_BUSY 25
#define RADIO_RESET 14
#define RADIO_TXEN 17
#define RADIO_RXEN 4 
#define INTERNAL_RADIO_CS 18

#define PACKET_SIZE 100

SX1262 radio = new Module(RADIO_CS, RADIO_DIO1, RADIO_RESET, RADIO_BUSY);

void setup() 
{
  pinMode(INTERNAL_RADIO_CS,OUTPUT);
  digitalWrite(INTERNAL_RADIO_CS,HIGH);
  Serial.begin(9600);
  while (!Serial && millis() < 10000);
  //Initialize radio module
  int state = radio.begin (  RADIO_FREQUENCY,
                             RADIO_BANDWIDTH,
                             RADIO_SF,
                             RADIO_CODE_RATE,
                             RADIO_SYNC_WORD,
                             RADIO_OUTPUT_POWER,
                             RADIO_PREAMBLE_LENGTH);
                             
  if (state == ERR_NONE) {
    Serial.println(F("success!"));
  } else {
    Serial.print(F("failed, code "));
    Serial.println(state);
    while (true);
  }
  // eByte E22-900M22S uses DIO3 to supply the external TCXO
  if (radio.setTCXO(1.8) == ERR_INVALID_TCXO_VOLTAGE)
  {
    Serial.println(F("Selected TCXO voltage is invalid for this module!"));
  }
  //RF switch pins setup
  radio.setRfSwitchPins(RADIO_RXEN, RADIO_TXEN);
}

void loop() 
{
  byte byteArr[PACKET_SIZE];
  int8_t tx_power = 0;
  while(tx_power <= 22)
  {
    radio.setOutputPower(tx_power);
    Serial.print(F("Transmitting packet @")); Serial.print(tx_power); Serial.print(" dBm... ");
    int state = radio.transmit(byteArr, PACKET_SIZE);
  
    if (state == ERR_NONE) {
      Serial.println(F("success!"));
    } else if (state == ERR_TX_TIMEOUT) {
      Serial.println(F("timeout!"));
    } else {
      Serial.print(F("failed, code "));
      Serial.println(state);
    }
    tx_power++;
    delay(5000);
  }
  Serial.println("Test ended");
  while(1);
}
```

The maximum output power i was able to get was 19.6dBm, when it was set to 13dBm in code. If i set the output power in the code anywhere from 13 to 22 dBm the measured power stays at 19.6dBm.

Maximum measured power in spectrum analyzer:
![image](https://user-images.githubusercontent.com/42620882/156574751-eae5c265-e1b3-452a-93f8-eee2f2433b61.png)

Also the max radio module current consumption while transmitting was 450mA while according to the datasheet it is supposed to be 650mA.

I have tried two different modules and the results were the same. The radio module was tested with an ESP32 and also with a SAMD51

I was wondering if someone already had this issue. Any kind of help is appreciated, thanks in advance!

**Comment by jgromes** [comment, marked as answer]
It doesn't seem like you changed the current limit in your code, so it's most likely at 60 mA (for the SX1262). It can be increased up to 140 mA. Regardles, I don't think that's the cause since you should be able to get the total +30 dBm output with very low SX1262 output power.

The YP2233W datasheet doesn't actually show the performance 915 MHz band, only 870 - 890 MHz and then 1.2 - 1.3 GHz. The information that YP2233W is the PA used in the Ebyte module is something I got from literally opening up one of the modules. It could have changed since then.

The weird thing is that when set to +4 dBm, you measured +10 dBm output, so there is some amplification going on, just a lot less. One more thing to try would be to just transmit the carrier with `transmitDirect`, that should get a more accurate measurement. Also I'm hoping there's a proper power supply in there.

I would suggest contacting Ebyte, to see if they have some more specification or verified example code.

**Reply by lucauriarte** [reply]
I forgot to mention that I was using an external 5V power source, and that I have also tried transmitting at 868 MHz and I got +1dBm compared to 915MHz.

I tried using transmitDirect and nothing seemed to change.

Changing the current limit solved the problem, I was able to get 29.43 dBm at 915MHz and 30.66dBm at 868MHz. I guess that the external PA is no longer the YP2233W, or for some reason it is not amplifying 26dB.

Here you can see the measurement at 868MHz, I was able to get that output power with an eternal power supply and also with batteries

![e22-900m30s max power 868](https://user-images.githubusercontent.com/42620882/157047617-5f1997dd-bc1a-485a-8236-171d876d3e25.jpg)

Thank you for helping, I really appreciate it!

**Reply by HeadBoffin** [reply]
> 30.66dBm at 868MHz

Depending on where you are, but just in case you aren't aware, that's not legal in the EU!

**Comment by danak6jq** [comment]
I've used both the E19 and E22 +30dBm modules. I built a test fixture and special firmware to automate power measurement.
These measurements were made using CW mode.

For the E22-900M30S, I found the following in a extract from an email to Doris at Ebyte:

My measurement of the E22-900M30S differs from what we've
discussed before. It appears that a reasonable adjustment for desired
TX output power >= +10dBm is to subtract 8dB. Below +10dBm, it
might be a bit more accurate to subtract 10dB.

Left: SX1262 power setting
Right: measured E22-915M30S output
All values in dBm

 -7:  5.9
 -6:  6.2
 -5:  6.8
 -4:  7.3
 -3:  7.6
 -2:  8.3
 -1:  8.8
  0:  9.6
  1:  10.4
  2:  11.1
  3:  11.7
  4:  12.8
  5:  13.5
  6:  14.6
  7:  15.5
  8:  16.5
  9:  17.5
 10:  18.4
 11:  19.3
 12:  20.3
 13:  21.3
 14:  22.2
 15:  23.1
 16:  24.6
 17:  25.7
 18:  26.5
 19:  27.7
 20:  28.5
 21:  29.1
 22:  29.6

Dana
